Provide jump to definition feature and auto complete

sid-kurias opened the pull request:
Works...for me :-)

sid-kurias committed:
Provide jump to definition feature.

sid-kurias committed:
Provide auto completion using company-mode.

sid-kurias committed:
Using bounds-of-thing-at-point to detect offset.

sid-kurias committed:
Looking at regions to determine where to navigate.

The analysis server returns a regions field which helps determine which
of the candidates is to be used for navigation.

sid-kurias committed:
fixed incorrect symbol length calculation.

nex3 commented on dart-mode.el:
```diff
@@ -737,6 +747,19 @@ the callback for that request is given the json decoded response."
+    	  (funcall (cdr resp-closure) msg))
+      (-if-let (err (assoc 'error msg))
+    	  (dart--analysis-server-on-error-callback msg)
+    	(dart-info (format "No callback was associated with id %s" raw-id)))))
```
Why are you duplicating all this code?

nex3 commented on dart-mode.el:
```diff
@@ -839,6 +862,40 @@ true for positions before the start of the statement, but on its line."
+(defun dart--process-nav-info (response offset)
+  "Report the possible completions and jump to the location of the definition.
+
+  Opens a new file in a new buffer if necessary."
```
Explain what the parameters are in the docstring.

Also, the second line should not be indented. You can run `checkdoc-defun` to lint your docstrings against standard Elisp style.

nex3 commented on dart-mode.el:
```diff
@@ -839,6 +862,40 @@ true for positions before the start of the statement, but on its line."
+  (-when-let* ((regions (cdr (assq 'regions (assq 'result response))))
+	       (target-offset
+		(loop for region across regions
+		      if (equal offset (cdr (assoc 'offset region)))
```
Nit: it's confusing that you're using `assoc` in some places and `assq` in others. I'd just use `assoc` everywhere, since that's what the rest of the library does.

nex3 commented on dart-mode.el:
```diff
@@ -839,6 +862,40 @@ true for positions before the start of the statement, but on its line."
+		(loop for region across regions
+		      if (equal offset (cdr (assoc 'offset region)))
+		      return (aref (cdr (assoc 'targets region)) 0)))
+	       (target (aref (cdr (assoc 'targets (assoc 'result response)))
```
Rather than writing `(assoc 'result response)` over and over, just assign it to a variable earlier in the let.

nex3 commented on dart-mode.el:
```diff
@@ -839,6 +862,40 @@ true for positions before the start of the statement, but on its line."
+		      return (aref (cdr (assoc 'targets region)) 0)))
+	       (target (aref (cdr (assoc 'targets (assoc 'result response)))
+			     target-offset))
+	       (files (aref (cdr (assoc 'files (assoc 'result response)))
```
Nit: call this variable `file`, because there's only one of it.

nex3 commented on dart-mode.el:
```diff
@@ -839,6 +862,40 @@ true for positions before the start of the statement, but on its line."
+			    (cdr (assoc 'fileIndex target))))
+	       (line (cdr (assoc 'startLine target)))
+	       (col  (cdr (assoc 'startColumn target))))
+    (find-file (format "%s" files))
```
Why are you using `format` here? The file should already be a string.

nex3 commented on dart-mode.el:
```diff
@@ -857,16 +914,16 @@ Key bindings:
-         "Cannot find `dart' executable or Dart analysis server snapshot.")
+	    (null dart-analysis-server-snapshot-path))
+	(dart-log
+	 "Cannot find `dart' executable or Dart analysis server snapshot.")
```
Please don't gratuitously change indentation.

sid-kurias committed:
refactored based on feedback received from Natalie

sid-kurias committed:
moved company-dart to its own repository

sid-kurias committed:
removed unwanted whitespace.

sid-kurias committed:
invalid reference to raw-id in execute callback

nex3 commented on dart-mode.el:
```diff
@@ -544,6 +545,15 @@ server sends a response to a request, it tags the response with the ID of the
 request.  We look up the callback for the request in this alist and run it with
 the JSON decoded server response.")
 
+(defvar dart--analysis-completion-callbacks  nil
```
Unless the IDs overlap (which I imagine they don't), I'd prefer to have all the callbacks in a single alist.

nex3 commented on dart-mode.el:
```diff
@@ -735,19 +745,33 @@ the callback for that request is given the json decoded response."
+
 (defun dart--analysis-server-handle-msg (msg)
   "Handle the parsed MSG from the analysis server."
+  (-when-let* ((event-assoc (assoc 'event msg))
```
Using two `-when-let*`s in a row here is confusing, because it looks like a message could be both a reply and an event. I'd use `if` to choose between the two `-when-let*`s.

sid-kurias committed:
Consolidate analysis callback lists into one.

Use one list to hold callbacks for completion events and normal
responses.
Using if/when for detecting whether an event or normal response is
received from the analysis server.

sid-kurias committed:
Refactored navigation to use subscriptions.

Refactored completion to detect completion.results event.

sid-kurias committed:
Refactored navigation to use subscriptions.

Refactored completion to detect completion.results event.

sid-kurias committed:
moved comment to appropriate location.

sid-kurias committed:
undoing unnecessary commit made.
